Update to 1.20.2

ellie-semele opened the pull request:
Been told https://github.com/QuiltMC/quilt-kotlin-libraries/pull/84 is urgent to merge otherwise built artifacts would be broken otherwise.

At the moment I've only updated versions to track with the latest FLK and get it to compile with `1.20.2`. There are some other changes which should likely be addressed before this gets merged:

- [x] The annotation `@CodecSerializable.Polymorphic` is not top level, `@MetaSerializable` used to allow this however it's now been deprecated and it only supports top level annotations.
- [x] There has been networking refactors and I or someone else probably need to add more extensions for those.
- [x] Parts of the code base is generated with Silver's math codegen project, those haven't been looked at but may need some updating.
- [x] Fabric Language Kotlin's language adapter seems to have broken due to changes in Kotlin 1.9.20, should verify our language adapter still works: [Changes for upstream since last bump](https://github.com/FabricMC/fabric-language-kotlin/compare/2c8819...HEAD)
- [x] Switch to using Yumi Licenser

sschr15 reviewed (commented):
It may be reasonable to split the 1.20.2 PR into a couple separate ones with prerelease builds to at least have a version available, but this looks pretty good

ellie-semele commented on library/src/main/kotlin/org/quiltmc/qkl/library/client/networking/ClientEvents.kt:
```diff
@@ -73,7 +74,7 @@ public typealias ClientPlayCallback = ClientPlayNetworkHandler.(
 
 public typealias ClientPlayJoinCallback = ClientPlayNetworkHandler.(
-    sender: PacketSender,
+    sender: PacketSender<CustomPayload>,
```
The event does specify the type:

```java
@ClientOnly
@FunctionalInterface
public interface Join extends ClientEventAwareListener {
	void onPlayReady(ClientPlayNetworkHandler handler, PacketSender<CustomPayload> sender, MinecraftClient client);
}```

ellie-semele commented on library/src/main/kotlin/org/quiltmc/qkl/library/brigadier/argument/IdentifierArguments.kt:
```diff
@@ -61,7 +61,7 @@ public fun ArgumentReader<
                 IdentifierArgumentType
                 >
-        >.asAdvancement(): Advancement {
+        >.asAdvancement(): C_unoypvme {
```
It's not mapped in 1.20.2 but it is mapped in later MC versions, will add a type alias.

ellie-semele commented on library/api/library.api:
No API methods have been removed, some did get renamed to match QSL and others had return types / parameter type changes to fit vanilla updates e.g. `Recipe` -> `RecipeHolder<Recipe>`.

If I need to bump the major version for the project I can do that though.

ellie-semele commented:
Can't see any changes to math types that are covered by QKL, switched to using `yumi-licenser` and fixed java sources not having their licenses checked.

ellie-semele commented:
Tried testing all entry point types the language adapter should provide and it seems like  2 are broken...

The following code won't work even though the same code being top level works just fine, it also doesn't work inside a companion object.

```kt
object ExampleMod {
    fun init(mod: ModContainer) {
        println("Hello Quilt from Example Mod!")
    }
}
```

ellie-semele commented:
I'd kindly request none of my contributions are used until I decide whether or not I want to contribute long term to this project.

ellie-semele commented:
tt
